src/medley/core.cljc

```diff
@@ -80,54 +90,66 @@
   "Maps a function over the key/value pairs of an associate collection. Expects
   a function that takes two arguments, the key and value, and returns the new
   key and value as a collection of two elements."
-  [f coll]
-  (reduce-map (fn [xf] (fn [m k v] (let [[k v] (f k v)] (xf m k v)))) coll))
+  ([f]
+   (completing-rf (fn [rf result [k v]] (rf result (f k v)))))
+  ([f coll]
+   (reduce-map (fn [xf] (fn [m k v] (let [[k v] (f k v)] (xf m k v)))) coll)))
 
 (defn map-keys
   "Maps a function over the keys of an associative collection."
-  [f coll]
-  (reduce-map (fn [xf] (fn [m k v] (xf m (f k) v))) coll))
+  ([f]
+   (completing-rf (fn [rf result [k v]] (rf result [(f k) v]))))
```

Comment:
A map entry should be used instead of a vector. We also need to check performance of destructuring vs. `key` and `val`.

Reply:
I forgot to mention this - I did do some benchmarks, and using plain vectors and destructuring turns out to be about 15% faster than `medley/map-entry`. 
Also the user-supplied function might return a vector, and using `key` and `val` would throw an exception.

Reply:
Ah, it's interesting that plain vectors are faster. We still need to ensure the result passed to the reducing function is a map entry though, as we want people to be able to use `key` and `val`.

Reply:
Another thing to note – I only did the benchmarks on Clojure using Criterium's quick-benchmark, and haven't done any tests on Clojurescript.